CoViewingExample/AudioDevices/ExampleAVPlayerProcessingTap.m

```diff
@@ -0,0 +1,373 @@
+                                                       UInt32 channelsOut) {
+    // Start with input buffer size as our argument.
+    // TODO: Does non-interleaving count towards the size (*2)?
+    // Give us a little more priming than we need (~8 frames).
+    UInt32 primeFrames = 8;
+    UInt32 sourceFrames = framesIn;
+    if (!isConverterPrimed) {
+        framesIn -= primeFrames;
+    } else if (*cachedSourceFrames > 0) {
+        framesIn += *cachedSourceFrames;
+    }
+    UInt32 desiredIoBufferSize = framesIn * 4 * bufferListIn->mNumberBuffers;
+//    printf("Input is %d bytes (%d total frames, %d cached frames).\n", desiredIoBufferSize, framesIn, *cachedSourceFrames);
+    UInt32 propertySizeIo = sizeof(desiredIoBufferSize);
+    AudioConverterGetProperty(converter,
+                              kAudioConverterPropertyCalculateOutputBufferSize,
+                              &propertySizeIo, &desiredIoBufferSize);
+
+    UInt32 bytesPerFrameOut = channelsOut * sizeof(SInt16);
+    UInt32 framesOut = (desiredIoBufferSize) / bytesPerFrameOut;
+//    UInt32 framesOut = (desiredIoBufferSize + (bytesPerFrameOut - 1)) / bytesPerFrameOut;
+//    framesOut += framesOut % 2;
+    UInt32 bytesOut = framesOut * bytesPerFrameOut;
+//    printf("Converter wants an output of %d bytes (%d frames, %d bytes per frames).\n",
+//           desiredIoBufferSize, framesOut, bytesPerFrameOut);
+
+    AudioBufferList *producerBufferList = TPCircularBufferPrepareEmptyAudioBufferList(buffer, 1, bytesOut, NULL);
+    if (producerBufferList == NULL) {
+        return;
+    }
+    producerBufferList->mBuffers[0].mNumberChannels = channelsOut;
+
+    OSStatus status;
+    UInt32 ioPacketSize = framesOut;
+//    printf("Ready to fill output buffer of frames: %d, bytes: %d with input buffer of frames: %d, bytes: %d.\n",
+//           framesOut, bytesOut, framesIn, framesIn * 4 * bufferListIn->mNumberBuffers);
+    ExampleAVPlayerAudioConverterContext context;
+    context.sourceBuffers = bufferListIn;
+    context.cacheBuffers = sourceCache;
+    context.sourcePackets = sourceFrames;
+    context.sourcePacketIndex = 0;
+    context.cachePackets = *cachedSourceFrames;
+    status = AudioConverterFillComplexBuffer(converter,
+                                             AVPlayerAudioTapConverterInputDataProc,
+                                             &context,
+                                             &ioPacketSize,
+                                             producerBufferList,
+                                             NULL);
+    // Adjust for what the format converter actually produced, in case it was different than what we asked for.
+    producerBufferList->mBuffers[0].mDataByteSize = ioPacketSize * bytesPerFrameOut;
+//    printf("Output was: %d packets / %d bytes. Consumed input packets: %d. Cached input packets: %d.\n",
+//           ioPacketSize, ioPacketSize * bytesPerFrameOut, context.sourcePackets, context.cachePackets);
```

Comment:
There are some commented code in this file but I guess you are aware of it and working on it.